script/prerender/manage.ts

```diff
@@ -0,0 +1,397 @@
+import { writeAssetFile } from './fileUtils';
+
+const {
+  ARCHIVE_URL,
+  CODE_VERSION,
+  OS_WEB_URL,
+  REACT_APP_ARCHIVE_URL,
+  REACT_APP_OS_WEB_API_URL,
+  RELEASE_ID,
+} = config;
+
+// Increasing this too much can lead to connection issues and greater memory usage in the manager
+const MAX_CONCURRENT_BOOKS = 5;
+
+// The worker fleet is automatically terminated after this many seconds
+// This is insurance in case this process gets stuck or crashes without deleting the workers stack
+const PRERENDER_TIMEOUT_SECONDS = 1800;
+
+// Abort the build if the workers stack is not ready after this many seconds
+const WORKERS_DEPLOY_TIMEOUT_SECONDS = 120;
+
+const BUCKET_NAME = process.env.BUCKET_NAME || 'sandbox-unified-web-primary';
+const BUCKET_REGION = process.env.BUCKET_REGION || 'us-east-1';
+const PUBLIC_URL = process.env.PUBLIC_URL || `/rex/releases/${RELEASE_ID}`;
+const RELEASE_ID_SUFFIX = RELEASE_ID.split('/', 2)[1];
+const WORK_REGION = process.env.WORK_REGION || 'us-east-2';
+const WORKERS_STACK_NAME = `rex-${CODE_VERSION}-${RELEASE_ID_SUFFIX}-prerender-workers`;
+
+const cfnClient = new CloudFormationClient({ region: WORK_REGION });
+const sqsClient = new SQSClient({ region: WORK_REGION });
+
+let archiveLoader: ReturnType<typeof createArchiveLoader>;
+let osWebLoader: ReturnType<typeof createOSWebLoader>;
+let bookLoader: ReturnType<typeof makeUnifiedBookLoader>;
+
+let resolveWorkQueuePromise: (workQueueUrl: string) => void;
+const workQueuePromise = new Promise<string>((resolve) => { resolveWorkQueuePromise = resolve; });
+
+let timeoutDate: Date;
+let numBooks = 0;
+let numPages = 0;
+
+async function renderManifest() {
+  writeAssetFile('/rex/release.json', JSON.stringify({
+    books: BOOKS,
+    code: CODE_VERSION,
+    id: RELEASE_ID,
+  }, null, 2));
+
+  writeAssetFile('/rex/config.json', JSON.stringify(config, null, 2));
+}
+
+// These functions begin the stack creation/deletion but do not wait for them to finish,
+// since we have other tasks to do while they are running
+async function createWorkersStack() {
+  // Set the timeoutDate, used also by manage()
+  timeoutDate = new Date(1000 * PRERENDER_TIMEOUT_SECONDS + new Date().getTime());
+
+  console.log(`Prerender timeout set to ${timeoutDate}`);
+  console.log('Started workers stack creation');
+
+  return cfnClient.send(new CreateStackCommand({
+    Parameters: [
+      {
+        ParameterKey: 'BucketName',
+        ParameterValue: BUCKET_NAME,
+      },
+      {
+        ParameterKey: 'BucketRegion',
+        ParameterValue: BUCKET_REGION,
+      },
+      {
+        ParameterKey: 'CodeVersion',
+        ParameterValue: CODE_VERSION,
+      },
+      {
+        ParameterKey: 'PublicUrl',
+        ParameterValue: PUBLIC_URL,
+      },
+      {
+        ParameterKey: 'ReleaseIdSuffix',
+        ParameterValue: RELEASE_ID_SUFFIX,
+      },
+      {
+        ParameterKey: 'ValidUntil',
+        ParameterValue: `${timeoutDate.toISOString().slice(0, -5)}Z`,
```

Comment:
what are you slicing off the end here? the zone? why?

Reply:
Have to slice off the milliseconds since AWS only accepts seconds

Reply:
we have a date uitll library with a [format](https://date-fns.org/v2.28.0/docs/format) function that you could use to construct the right thing instead of mangling the ISO string

Reply:
It seems date-fns 1.x does not support printing ISO strings if your local time is not GMT.
However, I can update it to 2.x and add the date-fns-tz library.

Reply:
I added date-fns-tz to convert the Date to UTC. I had to update date-fns because date-fns-tz fails to import with date-fns 1.x